java/ql/src/semmle/code/java/security/XPath.qll

```diff
@@ -1,65 +1,56 @@
-class XPath extends RefType {
-  XPath() { this.hasQualifiedName("javax.xml.xpath", "XPath") }
-}
-
-/** A call to `XPath.evaluate` or `XPath.compile` */
-class XPathEvaluateOrCompile extends XPathSink {
-  XPathEvaluateOrCompile() {
-    exists(Method m | this.getMethod() = m and m.getDeclaringType() instanceof XPath |
-      m.hasName(["evaluate", "compile"])
-    )
+abstract class XPathInjectionSink extends DataFlow::Node { }
+
+/** CSV sink models representing methods susceptible to XPath Injection attacks. */
+private class DefaultXPathInjectionSinkModel extends SinkModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "javax.xml.xpath;XPath;true;evaluate;;;Argument[0];xpath",
+        "javax.xml.xpath;XPath;true;evaluateExpression;;;Argument[0];xpath",
+        "javax.xml.xpath;XPath;true;compile;;;Argument[0];xpath",
+        "org.dom4j;Node;true;selectObject;;;Argument[0];xpath",
+        "org.dom4j;Node;true;selectNodes;;;Argument[0..1];xpath",
+        "org.dom4j;Node;true;selectSingleNode;;;Argument[0];xpath",
+        "org.dom4j;Node;true;numberValueOf;;;Argument[0];xpath",
+        "org.dom4j;Node;true;valueOf;;;Argument[0];xpath",
+        "org.dom4j;Node;true;matches;;;Argument[0];xpath",
+        "org.dom4j;Node;true;createXPath;;;Argument[0];xpath",
+        "org.dom4j;DocumentFactory;true;createPattern;;;Argument[0];xpath",
+        "org.dom4j;DocumentFactory;true;createXPath;;;Argument[0];xpath",
+        "org.dom4j;DocumentFactory;true;createXPathFilter;;;Argument[0];xpath",
+        "org.dom4j;DocumentHelper;false;createPattern;;;Argument[0];xpath",
+        "org.dom4j;DocumentHelper;false;createXPath;;;Argument[0];xpath",
+        "org.dom4j;DocumentHelper;false;createXPathFilter;;;Argument[0];xpath",
+        "org.dom4j.tree;AbstractNode;true;createXPathFilter;;;Argument[0];xpath",
+        "org.dom4j.tree;AbstractNode;true;createPattern;;;Argument[0];xpath",
+        "org.dom4j.util;ProxyDocumentFactory;true;createPattern;;;Argument[0];xpath",
+        "org.dom4j.util;ProxyDocumentFactory;true;createXPath;;;Argument[0];xpath",
+        "org.dom4j.util;ProxyDocumentFactory;true;createXPathFilter;;;Argument[0];xpath"
+      ]
   }
-
-  override Expr getSink() { result = this.getArgument(0) }
 }
 
-/** Any class extending or implementing `org.dom4j.Node` */
-class Dom4JNode extends RefType {
-  Dom4JNode() {
-    exists(Interface node | node.hasQualifiedName("org.dom4j", "Node") |
-      this.extendsOrImplements*(node)
+/** A default sink representing methods susceptible to XPath Injection attacks. */
+private class DefaultXPathInjectionSink extends XPathInjectionSink {
+  DefaultXPathInjectionSink() {
+    sinkNode(this, "xpath")
+    or
+    exists(ClassInstanceExpr constructor |
+      constructor.getConstructedType().getASourceSupertype*().hasQualifiedName("org.dom4j", "XPath")
+      or
+      constructor.getConstructedType().hasQualifiedName("org.dom4j.xpath", "XPathPattern")
```

Comment:
`XPathPattern` has a [constructor](https://javadoc.io/doc/org.dom4j/dom4j/latest/org/dom4j/xpath/XPathPattern.html#%3Cinit%3E(org.jaxen.pattern.Pattern)) taking a `org.jaxen.pattern.Pattern` argument, should that maybe be excluded?

Reply:
Well, that constructor will only produce an alert if the used `Pattern` is tainted by user input. Currently, since we are not adding additional taint steps from other `Pattern` constructors, we won't see it happening (unless `pattern` is another `XPathPattern`, in which case we'll catch the constructor call of _that_ `XPathPattern`). I added a test case in 0362e1c to demonstrate this.

But this leaves the door open for us to add new taint steps for `Pattern` constructors in the future if we consider it's needed, and in those cases an XPath expression will still be produced from tainted input (https://github.com/dom4j/dom4j/blob/master/src/main/java/org/dom4j/xpath/XPathPattern.java#L45-L62), so I think it's fair.

Having said that, we can discuss it further if you disagree.

Reply:
I was actually following the example of other test classes (e.g. https://github.com/github/codeql/blob/main/java/ql/test/query-tests/security/CWE-502/A.java) but now that you mention it, there are other test classes with descriptive names (e.g. https://github.com/github/codeql/blob/main/java/ql/test/query-tests/security/CWE-079/semmle/tests/XSS.java), so yes, probably it makes sense to improve the name.

Thanks for pointing this out! Fixed in aaabdce

Reply:
Thanks for the explanation. I am not completely familiar with how CodeQL taint tracking works and was afraid that if the Jaxen `Pattern` class is supported in the future, it would create two results: One for the creation of the `Pattern` and one for passing it to the `XPathPattern` constructor. But I am not sure if that would actually be the case.

It is probably fine as it is.